### PR TITLE
Add: test - error when there is no active connection

### LINKED ISSUE #156: error when there is no active connection
https://github.com/ploomber/jupysql/pull/126 passed all the tests but the docs failed: https://readthedocs.org/projects/jupysql/builds/19584152/

error:

```
    result = func(_payload, *args, **kwargs)
  File "/home/docs/checkouts/readthedocs.org/user_builds/jupysql/checkouts/latest/src/sql/plot.py", line 249, in boxplot
    ] = sql.connection.Connection.current._get_curr_connection_info()
AttributeError: 'NoneType' object has no attribute '_get_curr_connection_info'
looking for now-outdated files... none found
pickling environment... done
```

looks like an edge case when current is None, so we should ensure we only call `_get_curr_connection_info` when current is not None (and we should add a unit test to replicate this and ensure we catch it in the CI

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

Add test case to cover 

```
AttributeError: 'NoneType' object has no attribute '_get_curr_connection_info'
```

## Issue ticket number and link

[Closes #156](https://github.com/ploomber/jupysql/issues/156)

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--171.org.readthedocs.build/en/171/

### CODE REVIEW

File: src/tests/test_connection.py [hunk 1]
```diff
@@ -92,3 +97,8 @@ def test_missing_driver(
             Connection.from_connect_str(connect_str)
 
         assert "try to install package: " + missing_pkg in str(error.value)
+
+
+def test_no_current_connection_and_get_info(mock_init_connection):
```
Comment (edublancas) [inline src/tests/test_connection.py:102]: if we are only using mock_init_connection once we should define it inline here

File: src/tests/test_connection.py [hunk 2]
```diff
@@ -92,3 +97,8 @@ def test_missing_driver(
...
+    with pytest.raises(AttributeError):
+        Connection.current._get_curr_connection_info()
```
Comment (edublancas) [inline src/tests/test_connection.py:104]: shouldn't we be testing? that `Connection.current._get_curr_connection_info()` *doesnt fail* when current is None?

I think we have a leaky abstraction [here](https://github.com/ploomber/jupysql/blob/7b9f0f17c41342c31872708dd431633c7576b09b/src/sql/plot.py#L247), the if statement should be inside the `_get_curr_connection_info` and return None if there isn't an active connection. this way someone using the function doesn't have to know anything about the internals, as it's safe to call even if there are no connections

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/tests/test_connection.py (hunk 2 above)
Comment (tonykploomber) [reply src/tests/test_connection.py:104]: I am thinking to declare the `_get_curr_connection_info` as classmethod might make much sense

**Comment by tonykploomber** [comment]
@edublancas 
I set the `_get_curr_connection_info ` as classmethod.

Now when we invoke `_get_curr_connection_info`, the method itself will check if it has current connection associated with, if no, it will return `None`.

I think by doing this the consumer function doesn't have to know the `current` member in `Connection` class in order to invoke `_get_curr_connection_info`, which fixes the leaky abstraction issue


Now we can just trigger `sql.connection.Connection._get_curr_connection_info()` to get the info without exception even there is no current connection, [this is the test case](https://github.com/ploomber/jupysql/pull/171/files#diff-ecfdf2a1f3adc925d3c32b8aa52540474a10a4160d552f281d7dbd85a0260170R98)

File: src/sql/connection.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def get_missing_package_suggestion_str(e):
         module_name, MISSING_PACKAGE_LIST_EXCEPT_MATCHERS.keys()
     )
     if close_matches:
-        return "Perhaps you meant to use driver the dialect: \"{}\"".format(
+        return 'Perhaps you meant to use driver the dialect: "{}"'.format(
```
Comment (edublancas) [inline src/sql/connection.py:81]: cleaner version:

```
f"this is the value of something: {something!r}"
```

this ensures the quotes are added when printing the value in "something"

File: src/sql/connection.py [hunk 4]
```diff
@@ -337,6 +337,7 @@ def _close(cls, descriptor):
     def close(self):
         self.__class__._close(self)
 
+    @classmethod
```
Comment (edublancas) [inline src/sql/connection.py:340]: if this is a class method, the first argument should be cls, by convention

File: src/sql/connection.py [hunk 5]
```diff
@@ -78,9 +78,7 @@ def get_missing_package_suggestion_str(e):
         module_name, MISSING_PACKAGE_LIST_EXCEPT_MATCHERS.keys()
     )
     if close_matches:
```
Comment (idomic) [inline src/sql/connection.py:80]: Why do we need those `\` in here?

File: src/sql/magic.py [hunk 6]
```diff
@@ -296,7 +295,9 @@ def _execute(self, payload, line, cell, local_ns):
             creator=args.creator,
             alias=args.alias,
         )
-        payload["connection_info"] = conn._get_curr_connection_info()
+        payload[
+            "connection_info"
+        ] = sql.connection.Connection._get_curr_connection_info()
```
Comment (idomic) [inline src/sql/magic.py:300]: Why initialize two connections?

Comment (idomic) [reply src/sql/magic.py:300]: Like maybe we can to a check and initialize a single one.

Also seems like this block with `payload` is repetitive between `plot.py` and `magics.py`

**Review by idomic — COMMENTED** [review]
small stuff.

File: src/sql/connection.py (hunk 5 above)
Comment (tonykploomber) [reply src/sql/connection.py:80]: Cause we want to display the double quote around the dialect name
Ref: [Quotations](https://www.datacamp.com/tutorial/f-string-formatting-in-python#4.1.-quotations)

To show something like this:
```
Perhaps you meant to use driver the dialect: "sqlite"
```

File: src/sql/magic.py (hunk 6 above)
Comment (tonykploomber) [reply src/sql/magic.py:300]: Since this is the `payload["connection_info"]` here is for collecting `%sql [QUERY]`

However, in the `%sqlplot`, we still need to collect this info again since those are two command entry points

File: src/sql/connection.py (hunk 4 above)
Comment (tonykploomber) [reply src/sql/connection.py:338]: Fix

File: src/sql/connection.py (hunk 3 above)
Comment (tonykploomber) [reply src/sql/connection.py]: Fix in [Add: f-string fix](https://github.com/ploomber/jupysql/pull/171/commits/993e9a11c98ac198440ca4b8d0cc1438617fdf51)

Still need to keep the \" since we need to display double quote, Ref: [Quotations](https://www.datacamp.com/tutorial/f-string-formatting-in-python#4.1.-quotations)

If uses `{something!r}`, it will present single quote which we don't want

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
I merged this but then I realized the base branch was set to `156-error-when-there-is-no-active-connection` i guess this was a mistake @tonykploomber ?

**Comment by tonykploomber** [comment]
Somehow it was set incorrectly.

I opened another, please merge that.
https://github.com/ploomber/jupysql/pull/189
